src/compare.rs

```diff
@@ -0,0 +1,506 @@
+//! Comparators.
+//!
+//! A comparator is any type that implements the [`Compare`](trait.Compare.html) trait,
+//! which imposes a [total order](https://en.wikipedia.org/wiki/Total_order). Its
+//! [`compare`](trait.Compare.html#tymethod.compare) method accepts two values, which may
+//! be of the same type or different types, and returns an ordering on them.
+//!
+//! Comparators are useful for parameterizing the behavior of sort methods and certain
+//! data structures.
+//!
+//! The most basic comparator is [`Natural`](struct.Natural.html), which simply delegates
+//! to the type's implementation of [`Ord`]
+//! (http://doc.rust-lang.org/std/cmp/trait.Ord.html):
+//!
+//! ```rust
+//! use collect::compare::{Compare, Natural};
+//!
+//! let a = &1u8;
+//! let b = &2u8;
+//!
+//! let cmp = Natural;
+//! assert_eq!(cmp.compare(a, b), Less);
+//! assert_eq!(cmp.compare(b, a), Greater);
+//! assert_eq!(cmp.compare(a, a), Equal);
+//! ```
+//!
+//! There are convenience methods for checking each of the six relations:
+//!
+//! ```rust
+//! use collect::compare::{Compare, Natural};
+//!
+//! let a = &1u8;
+//! let b = &2u8;
+//!
+//! let cmp = Natural;
+//! assert!(cmp.compares_lt(a, b));
+//! assert!(cmp.compares_le(a, b));
+//! assert!(cmp.compares_ge(b, a));
+//! assert!(cmp.compares_gt(b, a));
+//! assert!(cmp.compares_eq(a, a));
+//! assert!(cmp.compares_ne(a, b));
+//! ```
+//!
+//! The [`CompareExt`](trait.CompareExt.html) trait provides extension methods that
+//! consume a comparator to produce a new one with different behavior, similar to
+//! [iterator adaptors](http://doc.rust-lang.org/std/iter/trait.IteratorExt.html). For
+//! example, all comparators can be [reversed](trait.CompareExt.html#method.rev):
+//!
+//! ```rust
+//! use collect::compare::{Compare, CompareExt, Natural};
+//!
+//! let cmp = Natural.rev();
```

Comment:
I think I'm reasonably happy with the status quo. Making it public is an... interesting... idea, but something that can be done later in a back-compat way.